### PR TITLE
Provider dependency management

### PR DESCRIPTION
**Pull request by NWalker1208**
This PR adds documentation for how our team will approach dependencies between providers. The decisions I've made are fairly arbitrary, but based on my research, I feel they should at least be sufficient for what we need.

I encourage everyone to read or at least skim through the updated README.md in the providers directory. That is where I have documented most of my design decisions. Any documentation that was previously in `ExampleProvider` or `BaseProvider` has been moved into that file as well.

I ended up needing to create a set of custom `SimpleChangeNotifierProxyProvider` widgets to replace the `ChangeNotifierProxyProvider` widgets that come with the Provider package. The latter have a rather clunky interface that would have made the source code for our AppProviders widget a bit bloated. My code is largely copied from the Provider package, with the necessary modifications to make it easier to use.

### CODE REVIEW

File: togetherly/lib/providers/simple_change_notifier_proxy_provider.dart
Comment (athibau2) [inline togetherly/lib/providers/simple_change_notifier_proxy_provider.dart:1]: What is the purpose of having 4 variations of the same definition? It's unclear and undocumented what the purpose of 1, 2, 3, and 4 are.

**Review by athibau2 — COMMENTED** [review]
Everything mostly looks good. The proxy provider file is hard to follow and understand, so I really have no clue what that is doing or its purpose. See comment.

File: togetherly/lib/providers/simple_change_notifier_proxy_provider.dart
Comment (NWalker1208) [reply togetherly/lib/providers/simple_change_notifier_proxy_provider.dart:1]: This is to mirror the family of `ChangeNotifierProxyProvider` classes  from the `provider` package (though they go all the way up to 6). Their reason, which is the same as mine, is that some providers need to depend on multiple other providers, and there is no good way with generic classes to accept an arbitrary number of type parameters. So instead, they provide classes for depending on 1 through 6 other providers, and if you need more than that, then you have to resort to using the 0-variant, which requires you to get each of the providers you need manually. See [this documentation](https://pub.dev/documentation/provider/latest/provider/ChangeNotifierProxyProvider2-class.html) for one of the variants from the library.

Comment (NWalker1208) [reply togetherly/lib/providers/simple_change_notifier_proxy_provider.dart:1]: I can push some documentation comments to make this clearer.

**Review by athibau2 — APPROVED** [review]
